### PR TITLE
`azurerm_virtual_machine_extension` - auto start and restore vm if needed

### LINKED ISSUE #17262: Support for automatically turning on VM w/ an extension so it can be deleted
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

As proposed here: https://github.com/hashicorp/terraform-provider-azurerm/issues/6098#issuecomment-1156049806, conditionally start a VM so it's possible to add/remove VM extensions, and then return it to it's original running state.

My use case is specifically so a VM can be deleted, but as suggested, this proposal is for the wider case.

### New or Affected Resource(s)/Data Source(s)

azurerm_virtual_machine_extension

### Potential Terraform Configuration

```hcl
resource "azurerm_virtual_machine_extension" "AADSSHLoginForLinux" {
  name                 = "AADSSHLoginForLinux"
  virtual_machine_id   = azurerm_linux_virtual_machine.workstation.id
  publisher            = "Microsoft.Azure.ActiveDirectory"
  type                 = "AADSSHLoginForLinux"
  type_handler_version = "1.0"
  auto_startstop_vm_to_modify_or_delete = true # <-- something wonderfully hacky like this?
}
```

### References

https://github.com/hashicorp/terraform-provider-azurerm/issues/6098

### PR DESCRIPTION
**Pull request by myc2h6o**
Close #17262

### CODE REVIEW

**Comment by myc2h6o** [comment]
Test result
image

File: internal/services/compute/virtual_machine_extension_resource.go [hunk 1]
```diff
@@ -182,6 +186,31 @@ func resourceVirtualMachineExtensionsCreateUpdate(d *pluginsdk.ResourceData, met
...
+		}
+		return fmt.Errorf("starting %q to create/update extension: %+v", *virtualMachineId, err)
+	}
+
+	// unlock the Virtual Machine ID immediately if power state is not changed, which indicates it is in running state at the first place, so that multiple extensions could be operated at the same time
+	// if Virtual Machine is stopped or deallocated, release the lock at the end, and multiple extensions are operated one at a time
+	if isPowerStateChanged {
+		defer locks.UnlockByID(virtualMachineId.ID())
+	} else {
+		locks.UnlockByID(virtualMachineId.ID())
```
Comment (ms-zhenhua) [inline internal/services/compute/virtual_machine_extension_resource.go:211]: If unlock immediately, will state change when running the following `CreateOrUpdate` or `WaitForCompletionRef`?

Comment (myc2h6o) [reply internal/services/compute/virtual_machine_extension_resource.go:211]: the provider won't change the state later if it unlocks immediately, as in this case all the different extensions will see the VM running and don't change its state.

**Comment by jackofallops** [comment]
Hi @myc2h6o - We've discussed this internally, and we feel like there's sufficient risk to starting/restoring VM's without user knowledge that this needs to be implemented in such a way as to require explicit opt in. Specifically, Terraform cannot know why the VM is in a stopped state, meaning we could potentially introduce undesired effects into a user's environment. 

I'm going to close this PR for now, but we're going to look to implement something that can facilitate this with explicit opt-in to the functionality.

Thanks again.

**Pull request closed without merging**